exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/UnionFunctions.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.drill.exec.expr.holders.UnionHolder;
+import org.apache.drill.exec.expr.holders.IntHolder;
+import org.apache.drill.exec.expr.holders.VarCharHolder;
+import org.apache.drill.exec.vector.complex.impl.UnionReader;
+import org.apache.drill.exec.vector.complex.reader.FieldReader;
+
+import javax.inject.Inject;
+
+public class UnionFunctions {
+
+  @FunctionTemplate(names = {"typeOf"},
+          scope = FunctionTemplate.FunctionScope.SIMPLE,
+          nulls = NullHandling.INTERNAL)
+  public static class GetType implements DrillSimpleFunc {
+
+    @Param
+    FieldReader input;
+    @Output
+    VarCharHolder out;
+    @Inject
+    DrillBuf buf;
+
+    public void setup() {}
+
+    public void eval() {
+
+      byte[] type;
+      if (input.isSet()) {
+         type = input.getType().getMinorType().name().getBytes();
+      } else {
+        type = org.apache.drill.common.types.TypeProtos.MinorType.NULL.name().getBytes();
+      }
+      buf = buf.reallocIfNeeded(type.length);
+      buf.setBytes(0, type);
+      out.buffer = buf;
+      out.start = 0;
+      out.end = type.length;
+    }
+  }
+
+  @SuppressWarnings("unused")
+  @FunctionTemplate(names = {"castUNION", "castToUnion"}, scope = FunctionTemplate.FunctionScope.SIMPLE, nulls=NullHandling.NULL_IF_NULL)
+  public static class CastUnionToUnion implements DrillSimpleFunc{
+
+    @Param FieldReader in;
+    @Output
+    UnionHolder out;
+
+    public void setup() {}
+
+    public void eval() {
+      out.reader = in;
+      out.isSet = in.isSet() ? 1 : 0;
+    }
+  }
+
+  @SuppressWarnings("unused")
+  @FunctionTemplate(name = "asList", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls=NullHandling.INTERNAL)
+  public static class CastUnionList implements DrillSimpleFunc {
+
+    @Param UnionHolder in;
+    @Output UnionHolder out;
+
+    public void setup() {}
+
+    public void eval() {
+      if (in.isSet == 1) {
+        out.reader = in.reader;
+      } else {
+        out.isSet = 0;
+      }
+    }
+  }
+
+  @SuppressWarnings("unused")
+  @FunctionTemplate(name = "is_LIST", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls=NullHandling.INTERNAL)
+  public static class UnionIsList implements DrillSimpleFunc {
+
+    @Param UnionHolder in;
+    @Output BitHolder out;
+
+    public void setup() {}
+
+    public void eval() {
+      if (in.isSet == 1) {
+        out.value = in.getType().getMinorType() == org.apache.drill.common.types.TypeProtos.MinorType.LIST ? 1 : 0;
+      } else {
+        out.value = 0;
+      }
+    }
+  }
+
+  @SuppressWarnings("unused")
+  @FunctionTemplate(name = "asMap", scope = FunctionTemplate.FunctionScope.SIMPLE, nulls=NullHandling.INTERNAL)
+  public static class CastUnionMap implements DrillSimpleFunc {
+
+    @Param UnionHolder in;
+    @Output UnionHolder out;
+
+    public void setup() {}
+
+    public void eval() {
+      if (in.isSet == 1) {
+        out.reader = in.reader;
```

Comment:
This seems like it should do checking rather than just convert. Also, shouldn't this be assert_map to be consistent with other functions?